src/sortedarray.c

```diff
@@ -0,0 +1,244 @@
+#endif
+
+/**
+ * Find the first index where the cmp_func of the sortedarray returns 0. There
+ * must be an element in the range from b to e where cmp_func returns 0 to get
+ * expected results.
+ *
+ * @param sortedarray   The SortedArray.
+ * @param data          The value to compare to.
+ * @param b             The start of the range (inclusive).
+ * @param e             The end of the range (inclusive).
+ *
+ * @return The first index where cmp_func returns 0.
+ */
+static unsigned int sortedarray_first_index(SortedArray *sortedarray,
+                                            SortedArrayValue data,
+                                            unsigned int b, unsigned int e);
+
+/**
+ * Find the last index where the cmp_func of the sortedarray returns 0. There
+ * must be an element in the range from b to e which makes cmp_func return 0 to
+ * get expected results.
+ *
+ * @param sortedarray   The SortedArray.
+ * @param data          The data to compare to.
+ * @param b             The start of the range (inclusive).
+ * @param e             The end of the range (inclusive).
+ *
+ * @return The last index where cmp_func returns 0.
+ */
+static unsigned int sortedarray_last_index(SortedArray *sortedarray,
+                                           SortedArrayValue data,
+                                           unsigned int b, unsigned int e);
+
+SortedArray *sortedarray_new(unsigned int length,
+                             SortedArrayEqualFunc equ_func,
+                             SortedArrayCompareFunc cmp_func)
+{
+    //check input requirements
+    if (equ_func == NULL || cmp_func == NULL) {
+        return NULL;
+    }
+    
+    ArrayList *array = arraylist_new(length);
+
+    // on failure, return null
+    if (array == NULL) {
+        return NULL;
+    }
+
+    SortedArray *sortedarray = malloc(sizeof(SortedArray));
+    
+    //check for failure
+    if (sortedarray == NULL) {
+        arraylist_free(array);
+        return NULL;
+    }
+    
+    //init
+    sortedarray->array = array;
+    sortedarray->equ_func = equ_func;
+    sortedarray->cmp_func = cmp_func;
+    return sortedarray;
+}
+
+void sortedarray_free(SortedArray *sortedarray)
+{
+    if (sortedarray != NULL) {
+        arraylist_free(sortedarray->array);
+
+        free(sortedarray);
+    }
+}
+
+void sortedarray_remove(SortedArray *sortedarray, unsigned int index)
+{
+    //just remove, because sorted prop will not be violated
+    arraylist_remove(sortedarray->array, index);
+}
+
+void sortedarray_remove_range(SortedArray *sortedarray, unsigned int index,
+                              unsigned int length)
+{
+    //just remove, because sorted prop will not be violated
+    arraylist_remove_range(sortedarray->array, index, length);
+}
+
+int sortedarray_insert(SortedArray *sortedarray, SortedArrayValue data)
+{
+    //do a binary search like loop to find right position
+    unsigned int b = 0;                             //left search index
```

Comment:
These short variable names make the code hard to read, and I don't quite understand why you used 'b', 'e' and 'r'. Probably 'left' and 'right' would be more readable.